File: Structural/Adapter/java/Adapter.java
/**
 * @brief The Adapter class makes the Adaptee compatible with the Target interface.
 *
 * The Adapter implements the Target interface and translates calls from the client
 * to the Adaptee's methods. This enables the client to work seamlessly with Adaptee
 * without being aware of its specific implementation.
 */
public class Adapter implements Target {
	private final Adaptee adaptee; // Composition: Adapter holds a reference to Adaptee.

	/**
	 * @brief Constructs the Adapter with an Adaptee instance.
	 *
	 * @param adaptee The Adaptee instance to adapt.
	 * @throws IllegalArgumentException if the adaptee is null.
	 */
	public Adapter(Adaptee adaptee) {
		if (adaptee == null) {
			throw new IllegalArgumentException("Adaptee cannot be null");
		}
		this.adaptee = adaptee;
	}

	/**
	 * @brief Implements the Target interface by delegating to the Adaptee.
	 *
	 * This method translates the Target's `request()` call into a call
	 * to the Adaptee's `specificRequest()` method.
	 *
	 * @return A string response translated by the Adapter.
	 */
	@Override
	public String request() {
		return "Adapter: Translating request to Adaptee's specificRequest().\n" + adaptee.specificRequest();
	}
}

File: Structural/Adapter/python/adapter.py
from typing import Protocol

class Target(Protocol):
	"""
	Defines the domain-specific interface used by the client.
	"""
	def request(self) -> str:
		"""
		Handles the request in a format expected by the client.
		:return: A string response.
		"""
		pass

class Adaptee:
	"""
	Defines the existing interface with specific functionality that needs adaptation.
	"""
	def specific_request(self) -> str:
		"""
		Provides functionality that needs to be adapted.
		:return: A specific string response.
		"""
		return "Adaptee: Specific request called."

class Adapter(Target):
	"""
	Adapts the Adaptee to the Target interface.
	"""
	def __init__(self, adaptee: Adaptee):
		if not adaptee:
			raise ValueError("Adaptee cannot be None")
		self._adaptee = adaptee

	def request(self) -> str:
		"""
		Translates the Target's request to the Adaptee's specific_request.
		:return: A string response translated by the Adapter.
		"""
		return f"Adapter: Translating request to Adaptee's specific_request.\n{self._adaptee.specific_request()}"
